File: library/src/main/java/com/snail/swipemenulistview/SwipeMenuLayout.java
package com.snail.swipemenulistview;

import android.content.Context;
import android.support.v4.view.GestureDetectorCompat;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.GestureDetector.OnGestureListener;
import android.view.GestureDetector.SimpleOnGestureListener;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.Interpolator;
import android.widget.AbsListView;
import android.widget.FrameLayout;
import android.widget.OverScroller;

public class SwipeMenuLayout extends FrameLayout {

    private static final int STATE_CLOSE = 0;
    private static final int STATE_OPEN = 1;

    private int mSwipeDirection;

    private View mContentView;
    private SwipeMenuView mMenuView;
    private int mDownX;
    private int mDownY;
    private float lastMoveX;
    private float moveDis;
    private int state = STATE_CLOSE;
    private GestureDetectorCompat mGestureDetector;
    private boolean isFling;
    private int MIN_FLING = dp2px(5);
    private int MAX_VELOCITYX = dp2px(200);
    private OverScroller mOpenScroller;
    private OverScroller mCloseScroller;
    private int mBaseX;
    private int position;
    private Interpolator mCloseInterpolator;
    private Interpolator mOpenInterpolator;

    private boolean mSwipEnable = true;
    private boolean isClick;

    public SwipeMenuLayout(View contentView, SwipeMenuView menuView) {
        this(contentView, menuView, null, null);
    }

    public SwipeMenuLayout(View contentView, SwipeMenuView menuView, Interpolator closeInterpolator, Interpolator openInterpolator) {
        super(contentView.getContext());
        mCloseInterpolator = closeInterpolator;
        mOpenInterpolator = openInterpolator;
        mContentView = contentView;
        mMenuView = menuView;
        mMenuView.setLayout(this);
        init();
    }

    private SwipeMenuLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    private SwipeMenuLayout(Context context) {
        super(context);
    }

    public int getPosition() {
        return position;
    }

    public void setPosition(int position) {
        this.position = position;
        mMenuView.setPosition(position);
    }

    public void setSwipeDirection(int swipeDirection) {
        mSwipeDirection = swipeDirection;
    }

    private void init() {
        setLayoutParams(new AbsListView.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT));
        OnGestureListener mGestureListener = new SimpleOnGestureListener() {
            @Override
            public boolean onDown(MotionEvent e) {
                isFling = false;
                return true;
            }

            @Override
            public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
                if (Math.abs(velocityX) > MAX_VELOCITYX && Math.abs(e1.getX() - e2.getX()) > MIN_FLING) {
                    isFling = true;
                }                
                return super.onFling(e1, e2, velocityX, velocityY);
            }
        };
        mGestureDetector = new GestureDetectorCompat(getContext(), mGestureListener);

        if (mCloseInterpolator != null) {
            mCloseScroller = new OverScroller(getContext(), mCloseInterpolator);
        } else {
            mCloseScroller = new OverScroller(getContext());
        }
        if (mOpenInterpolator != null) {
            mOpenScroller = new OverScroller(getContext(), mOpenInterpolator);
        } else {
            mOpenScroller = new OverScroller(getContext());
        }

        LayoutParams contentParams = new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
        mContentView.setLayoutParams(contentParams);
        mMenuView.setLayoutParams(new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT));

        addView(mContentView);
        addView(mMenuView);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
    }

    public boolean onSwipe(MotionEvent event) {
        mGestureDetector.onTouchEvent(event);
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                mDownX = (int) event.getX();
                mDownY = (int) event.getY();
                isFling = false;
                isClick = true;
                break;
            case MotionEvent.ACTION_MOVE:
                moveDis = lastMoveX - event.getX();
                lastMoveX = event.getX();
                int dis = (int) (mDownX - event.getX());
                if (Math.abs(dis) > dp2px(2)) {
                    isClick = false;
                }                
                if (state == STATE_OPEN) {
                    dis += mMenuView.getWidth() * mSwipeDirection;
                }
                swipe(dis);
                break;
            case MotionEvent.ACTION_UP:
                if (isFling) {
                    if (Math.signum(moveDis) == mSwipeDirection) {
                        smoothOpenMenu();
                    } else {
                        smoothCloseMenu();
                        return false;
                    }
                } else if (isClick || Math.abs(mDownX - event.getX()) < Math.abs(mDownY - event.getY())) {
                    smoothCloseMenu();
                    return false;
                } else if ((Math.abs(mDownX - event.getX()) > (mMenuView.getWidth() / 2) && Math.signum(mDownX - event.getX()) == mSwipeDirection && !isOpen()) ||
                        (Math.abs(mDownX - event.getX()) < (mMenuView.getWidth() / 2) && Math.signum(mDownX - event.getX()) == -mSwipeDirection && isOpen()) ||
                        (isOpen() && (mMenuView.getLeft() == 0 || mMenuView.getRight() == mContentView.getWidth()))) {
                    smoothOpenMenu();
                } else {
                    smoothCloseMenu();
                    return false;
                }                
                break;
        }
        return true;
    }

    public boolean isOpen() {
        return state == STATE_OPEN;
    }
    
    private void swipe(int dis) {
        if (!mSwipEnable) {
            return;
        }
        if (Math.signum(dis) != mSwipeDirection) {
            dis = 0;
        } else if (Math.abs(dis) > mMenuView.getWidth()) {
            dis = mMenuView.getWidth() * mSwipeDirection;
        }

        mContentView.layout(-dis, mContentView.getTop(), mContentView.getWidth() - dis, getMeasuredHeight());

        if (mSwipeDirection == SwipeMenuListView.DIRECTION_LEFT) {
            mMenuView.layout(mContentView.getWidth() - dis, mMenuView.getTop(),
                    mContentView.getWidth() + mMenuView.getWidth() - dis, mMenuView.getBottom());
        } else {
            mMenuView.layout(-mMenuView.getWidth() - dis, mMenuView.getTop(), -dis, mMenuView.getBottom());
        }
    }

    @Override
    public void computeScroll() {
        if (state == STATE_OPEN) {
            if (mOpenScroller.computeScrollOffset()) {
                swipe(mOpenScroller.getCurrX() * mSwipeDirection);
                postInvalidate();
            }
        } else {
            if (mCloseScroller.computeScrollOffset()) {
                swipe((mBaseX - mCloseScroller.getCurrX()) * mSwipeDirection);
                postInvalidate();
            }
        }
    }

    public void smoothCloseMenu() {
        state = STATE_CLOSE;
        if (mSwipeDirection == SwipeMenuListView.DIRECTION_LEFT) {
            mBaseX = -mContentView.getLeft();
            mCloseScroller.startScroll(0, 0, mMenuView.getWidth(), 0, 400);
        } else {
            mBaseX = mMenuView.getRight();
            mCloseScroller.startScroll(0, 0, mMenuView.getWidth(), 0, 400);
        }
        postInvalidate();
    }

    public void smoothOpenMenu() {
        if (!mSwipEnable) {
            return;
        }
        state = STATE_OPEN;
        if (mSwipeDirection == SwipeMenuListView.DIRECTION_LEFT) {
            mOpenScroller.startScroll(-mContentView.getLeft(), 0, mMenuView.getWidth(), 0, 400);
        } else {
            mOpenScroller.startScroll(mContentView.getLeft(), 0, mMenuView.getWidth(), 0, 400);
        }
        postInvalidate();
    }

    public void closeMenu() {
        if (mCloseScroller.computeScrollOffset()) {
            mCloseScroller.abortAnimation();
        }
        state = STATE_CLOSE;
        swipe(0);
    }

    public void openMenu() {
        if (!mSwipEnable) {
            return;
        }
        state = STATE_OPEN;
        swipe(mMenuView.getWidth() * mSwipeDirection);
    }

    public View getContentView() {
        return mContentView;
    }

    public SwipeMenuView getMenuView() {
        return mMenuView;
    }

    private int dp2px(int dp) {
        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp,
                getContext().getResources().getDisplayMetrics());
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        mMenuView.measure(MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED), 
                MeasureSpec.makeMeasureSpec(getMeasuredHeight(), MeasureSpec.EXACTLY));
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        mContentView.layout(0, 0, getMeasuredWidth(), mContentView.getMeasuredHeight());
        if (mSwipeDirection == SwipeMenuListView.DIRECTION_LEFT) {
            mMenuView.layout(getMeasuredWidth(), 0, getMeasuredWidth() + mMenuView.getMeasuredWidth(),
                    mContentView.getMeasuredHeight());
        } else {
            mMenuView.layout(-mMenuView.getMeasuredWidth(), 0, 0, mContentView.getMeasuredHeight());
        }
    }

    public void setMenuHeight(int measuredHeight) {
        LayoutParams params = (LayoutParams) mMenuView.getLayoutParams();
        if (params.height != measuredHeight) {
            params.height = measuredHeight;
            mMenuView.setLayoutParams(mMenuView.getLayoutParams());
        }
    }

    public void setSwipEnable(boolean swipEnable) {
        mSwipEnable = swipEnable;
    }

    public boolean getSwipEnable() {
        return mSwipEnable;
    }
}

File: library/src/main/java/com/snail/swipemenulistview/SwipeMenuLayout.kt
package com.snail.swipemenulistview

import android.util.TypedValue
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.View
import android.view.animation.Interpolator
import android.widget.AbsListView
import android.widget.FrameLayout
import android.widget.OverScroller
import androidx.core.view.GestureDetectorCompat

class SwipeMenuLayout(internal var contentView: View, internal var menuView: SwipeMenuView, closeInterpolator: Interpolator? = null, openInterpolator: Interpolator? = null) : FrameLayout(contentView.context) {
    private var mSwipeDirection: Int = 0
    private var mDownX: Int = 0
    private var mDownY: Int = 0
    private var lastMoveX: Float = 0.toFloat()
    private var moveDis: Float = 0.toFloat()
    private var state = STATE_CLOSE
    private var mGestureDetector: GestureDetectorCompat? = null
    private var isFling: Boolean = false
    private val minFling = dp2px(5)
    private val maxVelocityx = dp2px(200)
    private var mOpenScroller: OverScroller? = null
    private var mCloseScroller: OverScroller? = null
    private var mBaseX: Int = 0
    var position: Int = 0
        set(position) {
            field = position
            menuView.position = position
        }
    private var mCloseInterpolator: Interpolator? = closeInterpolator
    private var mOpenInterpolator: Interpolator? = openInterpolator

    var swipEnable = true
    private var isClick: Boolean = false

    val isOpen: Boolean
        get() = state == STATE_OPEN

    init {
        this.menuView.setLayout(this)
        init()
    }

    fun setSwipeDirection(swipeDirection: Int) {
        mSwipeDirection = swipeDirection
    }

    private fun init() {
        layoutParams = AbsListView.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT)
        val mGestureListener = object : SimpleOnGestureListener() {
            override fun onDown(e: MotionEvent): Boolean {
                isFling = false
                return true
            }

            override fun onFling(e1: MotionEvent, e2: MotionEvent, velocityX: Float, velocityY: Float): Boolean {
                if (Math.abs(velocityX) > maxVelocityx && Math.abs(e1.x - e2.x) > minFling) {
                    isFling = true
                }
                return super.onFling(e1, e2, velocityX, velocityY)
            }
        }
        mGestureDetector = GestureDetectorCompat(context, mGestureListener)

        mCloseScroller = if (mCloseInterpolator != null) {
            OverScroller(context, mCloseInterpolator)
        } else {
            OverScroller(context)
        }
        mOpenScroller = if (mOpenInterpolator != null) {
            OverScroller(context, mOpenInterpolator)
        } else {
            OverScroller(context)
        }

        val contentParams = FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT)
        contentView.layoutParams = contentParams
        menuView.layoutParams = FrameLayout.LayoutParams(FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT)

        addView(contentView)
        addView(menuView)
    }

    fun onSwipe(event: MotionEvent): Boolean {
        mGestureDetector!!.onTouchEvent(event)
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                mDownX = event.x.toInt()
                mDownY = event.y.toInt()
                isFling = false
                isClick = true
            }
            MotionEvent.ACTION_MOVE -> {
                moveDis = lastMoveX - event.x
                lastMoveX = event.x
                var dis = (mDownX - event.x).toInt()
                if (Math.abs(dis) > dp2px(2)) {
                    isClick = false
                }
                if (state == STATE_OPEN) {
                    dis += menuView.width * mSwipeDirection
                }
                swipe(dis)
            }
            MotionEvent.ACTION_UP -> if (isFling) {
                if (Math.signum(moveDis) == mSwipeDirection.toFloat()) {
                    smoothOpenMenu()
                } else {
                    smoothCloseMenu()
                    return false
                }
            } else if (isClick || Math.abs(mDownX - event.x) < Math.abs(mDownY - event.y)) {
                smoothCloseMenu()
                return false
            } else if (Math.abs(mDownX - event.x) > menuView.width / 2 && Math.signum(mDownX - event.x) == mSwipeDirection.toFloat() && !isOpen ||
                    Math.abs(mDownX - event.x) < menuView.width / 2 && Math.signum(mDownX - event.x) == (-mSwipeDirection).toFloat() && isOpen ||
                    isOpen && (menuView.left == 0 || menuView.right == contentView.width)) {
                smoothOpenMenu()
            } else {
                smoothCloseMenu()
                return false
            }
        }
        return true
    }

    private fun swipe(dis: Int) {
        var distance = dis
        if (!swipEnable) {
            return
        }
        if (Math.signum(distance.toFloat()) != mSwipeDirection.toFloat()) {
            distance = 0
        } else if (Math.abs(distance) > menuView.width) {
            distance = menuView.width * mSwipeDirection
        }

        contentView.layout(-distance, contentView.top, contentView.width - distance, measuredHeight)

        if (mSwipeDirection == SwipeMenuListView.DIRECTION_RIGHT_TO_LEFT) {
            menuView.layout(contentView.width - distance, menuView.top, contentView.width + menuView.width - distance, menuView.bottom)
        } else {
            menuView.layout(-menuView.width - distance, menuView.top, -distance, menuView.bottom)
        }
    }

    override fun computeScroll() {
        if (state == STATE_OPEN) {
            if (mOpenScroller!!.computeScrollOffset()) {
                swipe(mOpenScroller!!.currX * mSwipeDirection)
                postInvalidate()
            }
        } else {
            if (mCloseScroller!!.computeScrollOffset()) {
                swipe((mBaseX - mCloseScroller!!.currX) * mSwipeDirection)
                postInvalidate()
            }
        }
    }

    fun smoothCloseMenu() {
        state = STATE_CLOSE
        if (mSwipeDirection == SwipeMenuListView.DIRECTION_RIGHT_TO_LEFT) {
            mBaseX = -contentView.left
            mCloseScroller!!.startScroll(0, 0, menuView.width, 0, 400)
        } else {
            mBaseX = menuView.right
            mCloseScroller!!.startScroll(0, 0, menuView.width, 0, 400)
        }
        postInvalidate()
    }

    fun smoothOpenMenu() {
        if (!swipEnable) {
            return
        }
        state = STATE_OPEN
        if (mSwipeDirection == SwipeMenuListView.DIRECTION_RIGHT_TO_LEFT) {
            mOpenScroller!!.startScroll(-contentView.left, 0, menuView.width, 0, 400)
        } else {
            mOpenScroller!!.startScroll(contentView.left, 0, menuView.width, 0, 400)
        }
        postInvalidate()
    }

    fun closeMenu() {
        if (mCloseScroller!!.computeScrollOffset()) {
            mCloseScroller!!.abortAnimation()
        }
        state = STATE_CLOSE
        swipe(0)
    }

    fun openMenu() {
        if (!swipEnable) {
            return
        }
        state = STATE_OPEN
        swipe(menuView.width * mSwipeDirection)
    }

    private fun dp2px(dp: Int): Int {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp.toFloat(),
                context.resources.displayMetrics).toInt()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        menuView.measure(View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
                View.MeasureSpec.makeMeasureSpec(measuredHeight, View.MeasureSpec.EXACTLY))
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        contentView.layout(0, 0, measuredWidth, contentView.measuredHeight)
        if (mSwipeDirection == SwipeMenuListView.DIRECTION_RIGHT_TO_LEFT) {
            menuView.layout(measuredWidth, 0, measuredWidth + menuView.measuredWidth, contentView.measuredHeight)
        } else {
            menuView.layout(-menuView.measuredWidth, 0, 0, contentView.measuredHeight)
        }
    }

    fun setMenuHeight(measuredHeight: Int) {
        val params = menuView.layoutParams as FrameLayout.LayoutParams
        if (params.height != measuredHeight) {
            params.height = measuredHeight
            menuView.layoutParams = menuView.layoutParams
        }
    }

    companion object {
        private const val STATE_CLOSE = 0
        private const val STATE_OPEN = 1
    }
}
